crypto/pem/pem_passwd.c

```diff
@@ -0,0 +1,309 @@
+# include <errno.h>
+
+// We support two types of terminal interface:
+// - termios for Linux/Unix
+// - WIN32 Console for Windows
+# if !defined(OPENSSL_WINDOWS)
+#  include <termios.h>
+#  define DEV_TTY "/dev/tty"
+#  define TTY_STRUCT             struct termios
+#  define TTY_FLAGS              c_lflag
+#  define TTY_get(tty,data)      tcgetattr(tty, data)
+#  define TTY_set(tty,data)      tcsetattr(tty, TCSANOW, data)
+# else /* OPENSSL_WINDOWS */
+#  include <windows.h>
+# endif
+
+# define NUM_SIG 32
+
+static volatile sig_atomic_t intr_signal;
+static struct CRYPTO_STATIC_MUTEX console_global_mutex = CRYPTO_STATIC_MUTEX_INIT;
+
+# if !defined(OPENSSL_WINDOWS)
+static struct sigaction savsig[NUM_SIG];
+# else
+static void (*savsig[NUM_SIG]) (int);
+# endif
+
+#if defined(OPENSSL_WINDOWS)
+  DWORD tty_orig, tty_new;
+#else
+  TTY_STRUCT tty_orig, tty_new;
+#endif
+
+FILE *tty_in, *tty_out;
+int is_a_tty;
+
+
+static void popsig(void) {
+# if !defined(OPENSSL_WINDOWS)
+    for (int i = 1; i < NUM_SIG; i++) {
+        if (i == SIGUSR1 || i == SIGUSR2 || i == SIGKILL) {
+            continue;
+        }
+        sigaction(i, &savsig[i], NULL);
+    }
+# else
+    signal(SIGABRT, savsig[SIGABRT]);
+    signal(SIGFPE, savsig[SIGFPE]);
+    signal(SIGILL, savsig[SIGILL]);
+    signal(SIGINT, savsig[SIGINT]);
+    signal(SIGSEGV, savsig[SIGSEGV]);
+    signal(SIGTERM, savsig[SIGTERM]);
+# endif
+}
+
+static void recsig(int signal) {
+    intr_signal = signal;
+}
+
+static int discard_line_remainder(FILE *in) {
+    char buf[5];
+
+    do {
+        if (!fgets(buf, 4, in)) {
+            return 0;
+        }
+    } while (strchr(buf, '\n') == NULL);
+    return 1;
+}
+
+/* Signal handling functions */
+static void pushsig(void) {
+# if !defined(OPENSSL_WINDOWS)
+    struct sigaction sa;
+    OPENSSL_cleanse(&sa, sizeof(sa));
+
+    sa.sa_handler = recsig;
+
+    for (int i = 1; i < NUM_SIG; i++) {
+        if (i == SIGUSR1 || i == SIGUSR2 || i == SIGKILL) {
+            continue;
+        }
+        sigaction(i, &sa, &savsig[i]);
+    }
+# else // Specific error codes for windows
+    savsig[SIGABRT] = signal(SIGABRT, recsig);
+    savsig[SIGFPE] = signal(SIGFPE, recsig);
+    savsig[SIGILL] = signal(SIGILL, recsig);
+    savsig[SIGINT] = signal(SIGINT, recsig);
+    savsig[SIGSEGV] = signal(SIGSEGV, recsig);
+    savsig[SIGTERM] = signal(SIGTERM, recsig);
+# endif
+
+// set SIGWINCH handler to default so our workflow is not
+// triggered on user resizing of window
+# if defined(SIGWINCH) && defined(SIG_DFL)
+    signal(SIGWINCH, SIG_DFL);
+# endif
+}
+
+/* Console management functions */
+void openssl_console_acquire_mutex(void) {
+    CRYPTO_STATIC_MUTEX_lock_write(&console_global_mutex);
+}
+
+void openssl_console_release_mutex(void) {
+    CRYPTO_STATIC_MUTEX_unlock_write(&console_global_mutex);
+}
+
+int openssl_console_open(void) {
+    is_a_tty = 1;
+    if (CRYPTO_STATIC_MUTEX_is_write_locked(&console_global_mutex) != 1) {
```

Comment:
Since this is an internal function and this is just a "sanity" check for the caller, this condition should be in an `assert`.